csi/volumeutils.py

```diff
@@ -1009,13 +1015,156 @@ def reload_glusterfs(volume):
     return True
 
 
-def mount_glusterfs(volume, mountpoint, is_client=False):
+# noqa # pylint: disable=too-many-locals
+# noqa # pylint: disable=too-many-statements
+# noqa # pylint: disable=too-few-public-methods
+class VolfileElement:
+    """Class to represent multiple elements within a volfile"""
+
+    def __init__(self, name):
+        self.name = name
+        self.type = None
+        self.options = {}
+        self.subvolumes = None
+
+
+class Volfile:
+    """Class with volfile utils methods"""
+
+    def __init__(self, volfile, elements):
+        self.volfile = volfile
+        self.elements = elements
+
+    @classmethod
+    def parse(cls, volfile):
+        """Parse volfile into multiple volfile elements"""
+
+        element = None
+        elements = []
+        with open(volfile) as volf:
+            for line in volf:
+                line = line.strip()
+                if line.startswith("volume "):
+                    element = VolfileElement(line.split()[-1])
+                elif line == "end-volume":
+                    elements.append(element)
+                    element = None
+                elif line.startswith("option "):
+                    _, name, value = line.split()
+                    element.options[name] = value
+                elif line.startswith("subvolumes "):
+                    element.subvolumes = line.split(" ", 1)[-1]
+                elif line.startswith("type "):
+                    element.type = line.split()[-1]
+        return Volfile(volfile, elements)
+
+
+    def update_options_by_type(self, opts):
+        """
+        Update existing storage-options with the options
+        passed by user through storage-class
+        """
+
+        opt = {}
+        for element in self.elements:
+            if opts.get(element.type, None) is not None:
```

Comment:
Replace below lines with

```
element.options.update(opts.get(element.type, {}))
```

Reply:
Ack to rest all suggestions.
Here if we update without checking if the key is present, the options will be added to all matched types.
Example:
If `protocol/client.remote-port: NNNN` or `protocol/client.volfile-key: XYZ` is given where `protocol/client` has many instances, these changes will be reflected in all those instances. So check for given key exist is important in my opinion.

```yaml
volume 09d204a6-4ba4-11ec-ac5e-0242ac110003-ta
    type protocol/client
    option transport.socket.read-fail-log false
    option remote-host tie-breaker.kadalu.io
    option remote-port 24007
    option volfile-key /storage-pool-1
    option remote-subvolume /mnt
end-volume
volume storage-pool-1-client-0
    type protocol/client
    option transport.socket.read-fail-log false
    option volfile-key /storage-pool-1
    option remote-subvolume /bricks/storage-pool-1/data/brick
    option remote-host server-storage-pool-1-0-0.storage-pool-1
end-volume
volume storage-pool-1-client-1
    type protocol/client
    option transport.socket.read-fail-log false
    option volfile-key /storage-pool-1
    option remote-subvolume /bricks/storage-pool-1/data/brick
    option remote-host server-storage-pool-1-1-0.storage-pool-1
end-volume```
Here `volfile-key` at 1st element superfluous.

Reply:
What are the options provided by the user? 

It looks correct to me(Also I am not able to find the difference between oneline version and expanded one, expanded code does not add a option if it is already exists?)

Some of the options are not for customization. For example, port, path, volfile-key etc. If name level customization is required then either use update by name or Add `glusterfs` option `--xlator-option=*-client-0.volfile-key=XYZ`

Reply:
The first element `volume 09d204a6-4ba4-11ec-ac5e-0242ac110003-ta` did not have the option `volfile-key`. Yes just using this as an example. Works fine elements are not repeated like here `protocol/client` is present 3 times, rest are only one.